core/src/main/scala/kafka/admin/TopicCommand.scala

```diff
@@ -190,39 +190,22 @@ object TopicCommand extends Logging {
-    val reportUnavailablePartitions = if (opts.options.has(opts.reportUnavailablePartitionsOpt)) true else false
-    val reportOverriddenConfigs = if (opts.options.has(opts.topicsWithOverridesOpt)) true else false
     val liveBrokers = zkUtils.getAllBrokersInCluster().map(_.id).toSet
+
+    val reportUnderReplicatedPartitions = opts.options.has(opts.reportUnderReplicatedPartitionsOpt)
+    val reportUnavailablePartitions = opts.options.has(opts.reportUnavailablePartitionsOpt)
+    val reportOverriddenConfigs = opts.options.has(opts.topicsWithOverridesOpt)
+
     for (topic <- topics) {
       zkUtils.getPartitionAssignmentForTopics(List(topic)).get(topic) match {
         case Some(topicPartitionAssignment) =>
-          val describeConfigs: Boolean = !reportUnavailablePartitions && !reportUnderReplicatedPartitions
-          val describePartitions: Boolean = !reportOverriddenConfigs
-          val sortedPartitions = topicPartitionAssignment.toList.sortWith((m1, m2) => m1._1 < m2._1)
-          if (describeConfigs) {
+          val sortedPartitionReplicas: List[(Int, scala.Seq[Int])] = topicPartitionAssignment.toList.sortWith((m1, m2) => m1._1 < m2._1)
+          if (!reportUnavailablePartitions && !reportUnderReplicatedPartitions) {
             val configs = AdminUtils.fetchEntityConfig(zkUtils, ConfigType.Topic, topic)
-            if (!reportOverriddenConfigs || configs.size() != 0) {
-              val numPartitions = topicPartitionAssignment.size
-              val replicationFactor = topicPartitionAssignment.head._2.size
-              println("Topic:%s\tPartitionCount:%d\tReplicationFactor:%d\tConfigs:%s"
-                .format(topic, numPartitions, replicationFactor, configs.map(kv => kv._1 + "=" + kv._2).mkString(",")))
-            }
+            describeAndPrintConfig(topic, sortedPartitionReplicas, configs)
           }
-          if (describePartitions) {
-            for ((partitionId, assignedReplicas) <- sortedPartitions) {
-              val inSyncReplicas = zkUtils.getInSyncReplicasForPartition(topic, partitionId)
-              val leader = zkUtils.getLeaderForPartition(topic, partitionId)
-              if ((!reportUnderReplicatedPartitions && !reportUnavailablePartitions) ||
-                  (reportUnderReplicatedPartitions && inSyncReplicas.size < assignedReplicas.size) ||
-                  (reportUnavailablePartitions && (!leader.isDefined || !liveBrokers.contains(leader.get)))) {
-                print("\tTopic: " + topic)
-                print("\tPartition: " + partitionId)
-                print("\tLeader: " + (if(leader.isDefined) leader.get else "none"))
-                print("\tReplicas: " + assignedReplicas.mkString(","))
-                println("\tIsr: " + inSyncReplicas.mkString(","))
-              }
-            }
+          if (!reportOverriddenConfigs) {
+            describePartitions(zkUtils, liveBrokers, reportUnderReplicatedPartitions, topic, topicPartitionAssignment, sortedPartitionReplicas)
```

Comment:
I can read the code, I just don't get why is it so complex :)

The way I see it, there are two unrelated decisions:
1. Which topics to print (all topics, filtered topics, unavailable topics, under-replicated topics)
2. Whether or not to print configs

Right now the logic is a bit intermixed and seems to be all over the place. I hope it can be streamlined.